Major overhaul of mbstring (part 14)

alexdowad opened the pull request:
A bugfix and performance boost for `mb_detect_encoding` are included. I would like to have thorough unit tests for that function (before trying to improve the heuristics and optimize for performance), but don't have a corpus of texts in various legacy encodings readily available.

FYA @nikic

nikic commented on ext/mbstring/libmbfl/mbfl/mbfilter.c:
```diff
@@ -245,12 +245,12 @@ static int mbfl_estimate_encoding_likelihood(int c, void *void_data)
 		 * the candidate encoding is not the right one. */
 		data->score += 10;
-	} else if (php_unicode_is_punct(c)) {
+	} else if ((c >= 0x21 && c <= 0x2F) || (c >= 0x3A && c <= 0x40) || (c >= 0x2000 && c <= 0x206F)) {
```
These character ranges aren't equivalent to the previous check -- what's the rationale for checking these ranges, but not others?

alexdowad replied:
First of all, the rationale for checking for punctuation in the first place...

Some legacy encodings deliberately encode multi-byte characters using byte values which are in the ASCII range (like all the ISO-2022 encodings, HZ, etc). If such an encoding is misinterpreted as ASCII, then typically the mangled string will have a much higher density of punctuation characters than would normally be the case. Hence, our main interest here is in punctuation characters which are present in ASCII.

I've added another range of codepoints which is also used for punctuation, but not for any reason in particular. And we could certainly add still more ranges, though I don't know if it will really help anything. In any case, I do want to keep the number of ranges checked reasonably small, for performance.

Remember these are just heuristics. If we can find other heuristics which are fast, and which make good guesses most of the time, that would be great.

At the end of the day, what we *really* need is a good corpus of realistic example texts; then we could benchmark various candidate heuristics for detection accuracy and speed, and pick the best combination based on actual data.

nikic replied:
Okay, that makes sense. I think it would be useful to add that explanation as a comment. Maybe we can drop the non-ascii range if we don't have an immediate use for it.

nikic reviewed (commented):
Can you please split off the mb_detect_encoding changes and target PHP-8.1 for them?

alexdowad commented:
> Can you please split off the mb_detect_encoding changes and target PHP-8.1 for them?

Is that so the bugfix will be backported to PHP-8.1?

nikic commented:
> > Can you please split off the mb_detect_encoding changes and target PHP-8.1 for them?
> 
> Is that so the bugfix will be backported to PHP-8.1?

We land changes on the lowest branch they should go, and then merge them upwards into master.

alexdowad commented:
Updated base branch for this PR to PHP-8.1, which created a lot of junk in the GitHub conversation view. Rebased on PHP-8.1.

@nikic Please let me know if that's not what you were suggesting.

nikic commented on ext/mbstring/mbstring.c:
```diff
@@ -1699,6 +1699,11 @@ PHP_FUNCTION(mb_str_split)
 
+	if (ZSTR_LEN(str) == 0) {
+		array_init_size(return_value, 0);
+		return;
```
RETURN_EMPTY_ARRAY() is preferred.

alexdowad replied:
Fixed. Thank you!

nikic reviewed (approved):
Looks good, with the caveat that not all changes here can land on PHP-8.1 due to ABI breaks. I'd suggest to land the mb_detect_encoding changes on PHP-8.1 and the rest on master.

alexdowad commented:
> Looks good, with the caveat that not all changes here can land on PHP-8.1 due to ABI breaks. I'd suggest to land the mb_detect_encoding changes on PHP-8.1 and the rest on master.

@nikic, thanks (as always) for the great code review!

May I ask? Is the problem with Travis getting stuck when running the test suite on ARM64 a known issue? Or is it just happening to me? If it's just happening to me, I definitely will not land any commits until I figure out why it's happening.

nikic commented:
> May I ask? Is the problem with Travis getting stuck when running the test suite on ARM64 a known issue? Or is it just happening to me? If it's just happening to me, I definitely will not land any commits until I figure out why it's happening.

Yeah, that's a known issue :)

alexdowad commented:
> > May I ask? Is the problem with Travis getting stuck when running the test suite on ARM64 a known issue? Or is it just happening to me? If it's just happening to me, I definitely will not land any commits until I figure out why it's happening.
> 
> Yeah, that's a known issue :)

Would you like me to try to debug it? Or is the cause already known?

alexdowad commented:
OK, merged via CLI (since some of these commits were for `PHP-8.1` and some were for `master`).
